### DISCUSSION TITLE
Unclear how to fetch configured profile options

### DISCUSSION THREAD

**Discussion by jszwedko** [smithy-lang/smithy-rs#1354] [Q&A] [answered]
Hi all,

I'm attempting to load the AWS profiles to fetch configured options from them via [`ProfileSet#get`](https://docs.rs/aws-config/0.10.1/aws_config/profile/struct.ProfileSet.html#method.get). The only way I can see to load the `ProfileSet` is via [aws_config](https://docs.rs/aws-config/0.10.1/aws_config/index.html)::[profile](https://docs.rs/aws-config/0.10.1/aws_config/profile/index.html)::[load](https://docs.rs/aws-config/0.10.1/aws_config/profile/fn.load.html#) but the function takes unexpected types as arguments so I'm not sure how to call it. Am I missing something obvious?

Thank you!

**Comment by rcoh** [comment, marked as answer]
You can use `Fs::real()` and `Env::real()` — they're doc-hidden because they're internal APIs that mostly exist for testing. Can you give me more info about why you need to read the AWS profile directly though? Generally, it would all be exposed through aws-config and configured automatically as part of configuration

**Reply by jszwedko** [reply]
Ah, thanks @rcoh ! I didn't realize it was just docs hidden, I was thinking they were unexported. I'll give that a shot. For reference, I'm trying to implement https://github.com/awslabs/aws-sdk-rust/issues/261 as a custom provider until the SDK has native support for it.

**Reply by jszwedko** [reply]
Also, thanks for the quick response! 😄

**Reply by rcoh** [reply]
oh I see! It _may_ be easier to actually just do it within the repo 😅  (because we already have parsers / models of the JSON format that is used). (And we'd happily accept / mentor, etc. on any PRs)

**Reply by rcoh** [reply]
The [IMDS credentials provider](https://github.com/awslabs/smithy-rs/blob/0592a1bad4577dec56954e675bf6aa25c65c1d34/aws/rust-runtime/aws-config/src/imds/credentials.rs#L74) would be a good template to start from because it's a simple-ish credentials provider that:
- reads config from the profile
- loads credentials serialized in JSON

**Reply by jszwedko** [reply]
Ah, yeah, that's true. I was thinking it'd take me a bit to build up the context to do it in-tree, but that provider is pretty good example. I'll give it a shot if you all don't mind mentoring it along 😄 .

**Reply by rcoh** [reply]
realized that IMDS isn't a great example because the profile usage is buried. https://github.com/awslabs/smithy-rs/blob/bd6b65bd6b49b6290c954ceb1ec5fafc15f6fa79/aws/rust-runtime/aws-config/src/profile/region.rs#L93 is a much narrower example of how to load a profileset

**Reply by jszwedko** [reply]
Thanks for the pointers @rcoh ! I started a PR here: https://github.com/awslabs/smithy-rs/pull/1356 . I'm curious what you think of the direction.
